File: 0131-palindrome-partitioning/0131-palindrome-partitioning.c
int isPalindrome(char* s, int start, int end) {
    while (start < end) {
        if (s[start++] != s[end--]) return 0;
    }
    return 1;
}

// Recursive function to find palindrome partitioning
void dfs(char* s, int start, char*** result, int* returnSize, int** returnColumnSizes, char** path, int pathSize) {
    if (s[start] == '\0') { // Base case: reach the end of the string
        result[*returnSize] = (char**)malloc(sizeof(char*) * pathSize);
        (*returnColumnSizes)[*returnSize] = pathSize;
        for (int i = 0; i < pathSize; ++i) {
            result[*returnSize][i] = (char*)malloc(strlen(path[i]) + 1);
            strcpy(result[*returnSize][i], path[i]);
        }
        (*returnSize)++;
        return;
    }
    for (int i = start; s[i] != '\0'; ++i) {
        if (isPalindrome(s, start, i)) {
            // If s[start...i] is a palindrome, add it to the path and recurse
            path[pathSize] = (char*)malloc(i - start + 2);
            strncpy(path[pathSize], s + start, i - start + 1);
            path[pathSize][i - start + 1] = '\0'; // Null-terminate the new string
            dfs(s, i + 1, result, returnSize, returnColumnSizes, path, pathSize + 1);
            free(path[pathSize]); // Backtrack
        }
    }
}

char*** partition(char* s, int* returnSize, int** returnColumnSizes) {
    int len = strlen(s);
    char*** result = (char***)malloc(sizeof(char**) * pow(2, len)); // Assuming max 1000 partitions
    *returnColumnSizes = (int*)malloc(sizeof(int) * pow(2, len));
    *returnSize = 0;
    char** path = (char**)malloc(sizeof(char*) * len); // Maximum possible partition length is len
    dfs(s, 0, result, returnSize, returnColumnSizes, path, 0);
    free(path);
    return result;
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:

    def isPalindrome(self, s: str) -> bool:
        """
        주어진 문자열이 Palindrome인가?
        """
        return s == s[::-1]

    # 어떤 str에 대한 모든 분할을 담은 list return
    def getSublists(self, s: str, start: int, path: list, results: list) -> None:
        """
        Generate all palindromic partitions of s starting from 'start' and add them to 'results'.
        'path' keeps the current partition.
        """
        # If we've reached the end of the string, add the current partition to the results
        if start == len(s):
            results.append(path[:])
            return

        for end in range(start + 1, len(s) + 1):
            # If the current segment is a palindrome, recurse with this segment added to the path
            if self.isPalindrome(s[start:end]):
                self.getSublists(s, end, path + [s[start:end]], results)


    def partition(self, s: str) -> List[List[str]]:
        self.getSublists(s, 0, [], results:=[])
        return results